REGRESSION(291827@main): Web content is sometimes entirely unavailable to VoiceOver

pvollan opened the pull request:
#### 4d5c2620d80a91023b81e3f73b78ee667eed3337
<pre>
REGRESSION(291827@main): Web content is sometimes entirely unavailable to VoiceOver
<a href="https://bugs.webkit.org/show_bug.cgi?id=295537">https://bugs.webkit.org/show_bug.cgi?id=295537</a>
<a href="https://rdar.apple.com/155247558">rdar://155247558</a>

Reviewed by Chris Dumez.

When we get the first AX request in the UI process, we should initialize AX in all existing WebContent processes.
If we only initialize it for the one associated with the Web view that is active, AX will never be initialized in
the other WebContent processes.

* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/Shared/WebPageCreationParameters.serialization.in:
* Source/WebKit/Shared/WebProcessCreationParameters.h:
* Source/WebKit/Shared/WebProcessCreationParameters.serialization.in:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::initializeAccessibility): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::initializeAccessibility):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::enableAccessibilityIfNecessary):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::initializeAccessibility): Deleted.
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::initializeAccessibility):

Canonical link: <a href="https://commits.webkit.org/297202@main">https://commits.webkit.org/297202@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9380c951508dd689a8af770a56c77b9085182923

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110910 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30573 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21008 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116940 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/61178 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31253 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39159 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84329 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113858 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24971 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99865 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64773 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24323 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60738 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94358 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18068 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119742 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37955 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28207 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93285 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38331 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96138 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93110 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38160 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15903 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/33941 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17888 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37848 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43319 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37511 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40847 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39216 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
```diff
@@ -1766,6 +1766,15 @@ void addResult(HashMap<WebCore::FrameIdentifier, Ref<WebCore::LegacyWebArchive>>
+bool WebPageProxy::shouldInitializeNSAccessibility() const
+{
+#if PLATFORM(MAC)
+    return m_configuration->processPool().hasReceivedAXRequestInUIProcess() || NSWorkspace.sharedWorkspace.isVoiceOverEnabled;
```
This doesn't cover for the huge variety of other assistive technologies we have, right? Voice Control, Accessibility Inspector, Full Keyboard Access, Switch Control, more.

minorninth replied:
Isn't checking for AXEnhancedUserInterface the most comprehensive?

pvollan replied:
That is a good point. The latest patch is taking a more generic approach. Out of curiosity, would AXEnhancedUserInterface cover other assistive technologies?

Thanks for reviewing, all!

minorninth replied:
All of our assistive technology that needs the accessibility tree from WebKit sets AXEnhancedUserInterface, yes. That should cover VoiceOver, Voice Control, Hover Text, etc.

twilco replied:
What about 3rd party assistive technologies? Can we count on all assistive technologies setting `AXEnhancedUserInterface`?

minorninth replied:
There aren't any full-fledged ATs I'm aware of, but there are things like Grammarly. I'm not sure they set AXEnhancedUserInterface, but accessing accessibility attributes of the web page should also enable web accessibility, right?

cdumez commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -3738,8 +3748,15 @@ static String commandNameForSelector(SEL selector)
+        auto processes = processPool->processes();
+        for (auto& process : processes) {
+            for (auto& page : process->pages()) {
+                page->initializeAccessibility();
```
I don't think we need to go via pages here? How about we create a function on WebProcess pool which iterates over all its processes and initializes AX for each one of them. I think it'd be clearer and maybe more robust as well.

minorninth replied:
Also, how does this work with site isolation? If a process contains only an iframe does it still have a page?

cdumez replied:
With my proposed approach, I think it will work with site isolation I believe. This is what I meant by "more robust".

pvollan replied:
That is a good point, I have addressed this in the latest patch.

Thanks for reviewing, all!

cdumez commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -3738,8 +3738,7 @@ static String commandNameForSelector(SEL selector)
-        m_page->initializeAccessibility();
-        accessibilityRegisterUIProcessTokens();
+        processPool->initializeAccessibility();
         processPool->markHasReceivedAXRequestInUIProcess();
```
We don't need this function anymore. `initializeAccessibility()` can just set that flag I think?

cdumez replied:
I also think it may be nice if `initializeAccessibility()` was renamed `initializeAccessibilityIfNecessary()` and if it would check the `hasReceivedAXRequestInUIProcess` flag internally so call sites do not have to worry about it.

cdumez commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -2855,4 +2855,20 @@ String WebProcessPool::platformResourceMonitorRuleListSourceForTesting()
+    if (m_hasReceivedAXRequestInUIProcess)
+        return;
+
+    for (auto& process : m_processes) {
```
Please wait for the safer cpp bot in case we need to `Ref` here.

cdumez commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -3734,13 +3734,9 @@ static String commandNameForSelector(SEL selector)
-        accessibilityRegisterUIProcessTokens();
-        processPool->markHasReceivedAXRequestInUIProcess();
+    if (![attribute isEqualToString:NSAccessibilityParentAttribute] && ![attribute isEqualToString:NSAccessibilityPositionAttribute]) {
+        Ref processPool = m_page->protectedLegacyMainFrameProcess()->processPool();
```
I recommend we use this:
```
m_page->configuration().protectedProcessPool()->initializeAccessibilityIfNecessary();
``
Using `protectedLegacyMainFrameProcess()` in new code is a red-flag.

cdumez commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -1649,6 +1649,25 @@ static float currentBacklightLevel()
 }
 
+#if ENABLE(INITIALIZE_ACCESSIBILITY_ON_DEMAND)
+void WebProcess::initializeAccessibility(Vector<SandboxExtension::Handle>&& handles)
```
Do you think we could add an assertion to make sure this is never called more than once?

cdumez reviewed (commented):
Can you please move `hasReceivedAXRequestInUIProcess` out of WebPageCreationParameters too? It should be on the InitializationParameters from WebProcessProxy instead.

cdumez commented on Source/WebKit/UIProcess/WebProcessPool.h:
```diff
@@ -630,7 +630,6 @@ class WebProcessPool final
 #endif
 
-    void markHasReceivedAXRequestInUIProcess() { m_hasReceivedAXRequestInUIProcess = true; }
     bool hasReceivedAXRequestInUIProcess() const { return m_hasReceivedAXRequestInUIProcess; }
```
I think we can remove this getter.

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
```diff
@@ -538,6 +538,10 @@ static void logProcessPoolState(const WebProcessPool& pool)
 #endif
+
+#if ENABLE(INITIALIZE_ACCESSIBILITY_ON_DEMAND)
+    parameters.shouldInitializeAccessibility = hasReceivedAXRequestInUIProcess();
```
I think we can use `m_hasReceivedAXRequestInUIProcess` instead of `hasReceivedAXRequestInUIProcess()` here and drop the getter.

cdumez commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -1649,6 +1649,29 @@ static float currentBacklightLevel()
+#if ENABLE(INITIALIZE_ACCESSIBILITY_ON_DEMAND)
+void WebProcess::initializeAccessibility(Vector<SandboxExtension::Handle>&& handles)
+{
+    static bool hasInitializedAccessibility = false;
```
`#if ASSERT_ENABLED`

cdumez commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -1649,6 +1649,29 @@ static float currentBacklightLevel()
+    static bool hasInitializedAccessibility = false;
+    ASSERT_UNUSED(hasInitializedAccessibility, !hasInitializedAccessibility);
+    hasInitializedAccessibility = true;
+
```
`#endif`

cdumez reviewed (approved):
LGTM with nits.

pvollan committed:
REGRESSION(291827@main): Web content is sometimes entirely unavailable to VoiceOver
https://bugs.webkit.org/show_bug.cgi?id=295537
rdar://155247558

Reviewed by Chris Dumez.

When we get the first AX request in the UI process, we should initialize AX in all existing WebContent processes.
If we only initialize it for the one associated with the Web view that is active, AX will never be initialized in
the other WebContent processes.

* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/Shared/WebPageCreationParameters.serialization.in:
* Source/WebKit/Shared/WebProcessCreationParameters.h:
* Source/WebKit/Shared/WebProcessCreationParameters.serialization.in:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::initializeAccessibility): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::initializeAccessibility):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::enableAccessibilityIfNecessary):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::initializeAccessibility): Deleted.
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::initializeAccessibility):

Canonical link: https://commits.webkit.org/297202@main
